src/charm.py

```diff
@@ -0,0 +1,595 @@
+        unknown = 0
+        runner_names = []
+
+        runner_info = runner_manager.get_github_info()
+
+        for runner in runner_info:
+            if runner.status == GitHubRunnerStatus.ONLINE.value:
+                online += 1
+                runner_names.append(runner.name)
+            elif runner.status == GitHubRunnerStatus.OFFLINE.value:
+                offline += 1
+            else:
+                # might happen if runner dies and GH doesn't notice immediately
+                unknown += 1
+        event.set_results(
+            {
+                "online": online,
+                "offline": offline,
+                "unknown": unknown,
+                "runners": ", ".join(runner_names),
+            }
+        )
+
+    @catch_unexpected_action_errors
+    def _on_reconcile_runners_action(self, event: ActionEvent) -> None:
+        """Handle the action of reconcile of runner state.
+
+        Args:
+            event: Action event of reconciling the runner.
+        """
+        runner_manager = self._get_runner_manager()
+        if not runner_manager:
+            event.fail("Missing token or org/repo path config")
+            return
+
+        delta = self._reconcile_runners(runner_manager)
+
+        self._on_check_runners_action(event)
+        event.set_results(delta)
+
+    @catch_unexpected_action_errors
+    def _on_flush_runners_action(self, event: ActionEvent) -> None:
+        """Handle the action of flushing all runner and reconciling afterwards.
+
+        Args:
+            event: Action event of flushing all runners.
+        """
+        runner_manager = self._get_runner_manager()
+        if not runner_manager:
+            event.fail("Missing token or org/repo path config")
+            return
+
+        runner_manager.flush()
+        delta = self._reconcile_runners(runner_manager)
+
+        self._on_check_runners_action(event)
+        event.set_results(delta)
+
+    @catch_unexpected_charm_errors
+    def _on_stop(self, _: StopEvent) -> None:
+        """Handle the stopping of the charm.
+
+        Args:
+            event: Event of stopping the charm.
+        """
+        try:
+            self._event_timer.disable_event_timer("update-runner-bin")
+            self._event_timer.disable_event_timer("reconcile-runners")
+        except TimerDisableError as ex:
+            logger.exception("Failed to stop the timer")
+            self.unit.status = BlockedStatus(f"Failed to stop charm event timer: {ex}")
+
+        runner_manager = self._get_runner_manager()
+        if runner_manager:
+            try:
+                runner_manager.flush()
+            # Safe guard against unexpected error.
+            except Exception:  # pylint: disable=broad-exception-caught
+                # Log but ignore error since we're stopping anyway.
+                logger.exception("Failed to clear runners")
+
+    def _reconcile_runners(self, runner_manager: RunnerManager) -> Dict[str, "JsonObject"]:
```

Comment:
Minor nitpick, would this be better named something like `_get_runner_diff`? I was thinking if _reconcile_runners would be suggestive of taking the reconcile action. :D

Reply:
This action does trigger reconciliation of runners at line 480.